tiledb/sm/filter/float_scaling_filter.cc

```diff
@@ -0,0 +1,298 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class FloatScalingFilter.
+ */
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/enums/filter_option.h"
+#include "tiledb/sm/filter/filter_buffer.h"
+#include "tiledb/sm/tile/tile.h"
+
+#include "tiledb/sm/filter/float_scaling_filter.h"
+
+#include <cassert>
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+void FloatScalingFilter::dump(FILE* out) const {
+  if (out == nullptr)
+    out = stdout;
+  fprintf(
+      out,
+      "FloatScalingFilter: BIT_WIDTH=%u, SCALE=%lf, OFFSET=%lf",
+      static_cast<uint32_t>(bit_width_),
+      scale_,
+      offset_);
+}
+
+Status FloatScalingFilter::serialize_impl(Buffer* buff) const {
+  Metadata buffer_struct = {scale_, offset_, bit_width_};
+  RETURN_NOT_OK(buff->write(&buffer_struct, sizeof(Metadata)));
+  return Status::Ok();
+}
+
+template <typename T, typename W>
+Status FloatScalingFilter::run_forward(
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  auto input_parts = input->buffers();
+  uint32_t num_parts = static_cast<uint32_t>(input_parts.size());
+  uint32_t metadata_size = sizeof(uint32_t) + num_parts * sizeof(uint32_t);
+  RETURN_NOT_OK(output_metadata->append_view(input_metadata));
+  RETURN_NOT_OK(output_metadata->prepend_buffer(metadata_size));
+  RETURN_NOT_OK(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  for (auto& i : input_parts) {
+    uint32_t s = i.size();
+    assert(s % sizeof(T) == 0);
+    uint32_t num_elems_in_part = (s / sizeof(T));
+    uint32_t new_size = num_elems_in_part * sizeof(W);
+    RETURN_NOT_OK(output_metadata->write(&new_size, sizeof(uint32_t)));
+    RETURN_NOT_OK(output->prepend_buffer(new_size));
+
+    const T* part_data = static_cast<const T*>(i.data());
+    for (uint32_t j = 0; j < num_elems_in_part; ++j) {
+      T elem = part_data[j];
+      W converted_elem = static_cast<W>(
+          trunc((elem - static_cast<T>(offset_)) / static_cast<T>(scale_)));
```

Comment:
I think this should probably use `std::round` (closest integer) instead of `trunc` (round towards zero).

Reply:
I used truncate because that was the spec in the design document. Per the design doc, 
> Proposed formula:
> # given
> #  IntN: type selected by the user (Int8, Int16, Int32)
> #  scale: float64 selected by user
> #  offset: float64 selected by user
> 
> stored_int: IntN = truncate(IntN, (raw_float - offset) / scale)


@ihnorton should we switch it to round?

Reply:
Good point, yes - let's use round; trunc will be biased. (round matches NetCDF and GRIB, so it's what most users will expect)

Reply:
Cool cool, am using round now.